wasm/src/object.c

```diff
@@ -71,6 +74,251 @@ static void __copy_object_elem(opa_object_t *result, opa_value *a, opa_value *b)
+
+            opa_array_t *parts = opa_cast_array(s);
+
+            for (int i = 0; i < parts->len; i++)
+            {
+                opa_value *s = opa_strings_replace(parts->elems[i].v, opa_string_terminated("~1"), opa_string_terminated("/"));
+                opa_value *part = opa_strings_replace(s, opa_string_terminated("~0"), opa_string_terminated("~"));
+                opa_array_append(path_segments, part);
+            }
+
+            return path_segments;
+        }
+        case OPA_ARRAY:
+        {
+            opa_array_t *y = opa_cast_array(a);
+
+            for (int i = 0; i < y->len; i++)
+            {
+                opa_array_append(path_segments, y->elems[i].v);
+            }
+
+            return path_segments;
+        }
+    }
+
+    return NULL;
+}
+
+opa_array_t *__get_json_paths(opa_value *a)
+{
+    opa_array_t *paths = opa_cast_array(opa_array());
+
+    switch (opa_value_type(a))
+    {
+        case OPA_SET:
+        {
+            opa_set_t *x = opa_cast_set(a);
+
+            for (int i = 0; i < x->n; i++)
+            {
+                opa_set_elem_t *elem = x->buckets[i];
+
+                while (elem != NULL)
+                {
+                    opa_array_t* path = __parse_path(elem->v);
+
+                    if (path == NULL)
+                    {
+                       return NULL;
+                    }
+                    opa_array_append(paths, &path->hdr);
+                    elem = elem->next;
+                }
+            }
+
+            return paths;
+        }
+        case OPA_ARRAY:
+        {
+            opa_array_t *y = opa_cast_array(a);
+
+            for (int i = 0; i < y->len; i++)
+            {
+                opa_array_t* path = __parse_path(y->elems[i].v);
+
+                if (path == NULL)
+                {
+                    return NULL;
+                }
+                opa_array_append(paths, &path->hdr);
+            }
+
+            return paths;
+        }
+    }
+
+    return NULL;
+}
+
+opa_value *__json_remove(opa_value *a, opa_value *b)
+{
+    if (b == NULL)
+    {
+        // The paths diverged, return a
+        return a;
+    }
+
+    opa_value* bObj;
+    switch (opa_value_type(b))
+    {
+        case OPA_OBJECT:
+        {
+            bObj = b;
+            break;
+        }
+        case OPA_NULL:
+        {
+            // Means we hit a leaf node on "b", dont add the value for a
+            return NULL;
+        }
+        default:
+            // The paths diverged, return a
+            return a;
+    }
+
+    switch (opa_value_type(a))
+    {
+        case OPA_STRING:
+        case OPA_NUMBER:
+        case OPA_BOOLEAN:
+        case OPA_NULL:
+        {
+            return a;
+        }
+        case OPA_OBJECT:
+        {
+            opa_object_t *new_obj = opa_cast_object(opa_object());
+
+            for (opa_value *key = opa_value_iter(a, NULL); key != NULL;
+                key = opa_value_iter(a, key))
```

Comment:
nit: no need to split across multiple lines.

Reply:
Done